### PR TITLE
ZEN-29009 Rewrite NTPMonitor

### PR DESCRIPTION
**Pull request by karolzen**
https://jira.zenoss.com/browse/ZEN-29009

### CODE REVIEW

File: ZenPacks/zenoss/NtpMonitor/datasources/NtpMonitorDataSource.py [hunk 1]
```diff
@@ -1,114 +1,163 @@
...
+                output += "|offset=%.10gs;%.6f;%.6f;" % (
+                    result["offset"], datasource.params["warning"], datasource.params["critical"]
+                )
+                severity = ZenEventClasses.Clear
+                data["values"][None]["offset"] = result["offset"]
+
+        data["events"].append({
+            "eventKey": event_key,
+            "summary": output,
+            "message": output,
```
Comment (StevePC) [inline ZenPacks/zenoss/NtpMonitor/datasources/NtpMonitorDataSource.py]: You might want to put a smaller summary in place and then use output for the message.

File: ZenPacks/zenoss/NtpMonitor/datasources/NtpMonitorDataSource.py [hunk 2]
```diff
@@ -1,114 +1,163 @@
...
+                output += " Offset %.10g secs" % result["offset"]
+            if result["offset_result"] != STATE_UNKNOWN:
+                output += "|offset=%.10gs;%.6f;%.6f;" % (
+                    result["offset"], datasource.params["warning"], datasource.params["critical"]
+                )
+                severity = ZenEventClasses.Clear
+                data["values"][None]["offset"] = result["offset"]
+
+        data["events"].append({
+            "eventKey": event_key,
```
Comment (StevePC) [inline ZenPacks/zenoss/NtpMonitor/datasources/NtpMonitorDataSource.py]: You should generally have a different eventKey for errors involving connection and processing vs events generated due to the content of the result, as clearing one of them might not require clearing the other.  For instance, as written, if you generate an error event becuase you couldn't connect to NTP, then when the conenction starts working, if the offset is STATE_UNKNOWN, you won't clear the connection error event, even though it is now connecting.

File: ZenPacks/zenoss/NtpMonitor/datasources/NtpMonitorDataSource.py [hunk 3]
```diff
@@ -1,114 +1,163 @@
...
 
+log = logging.getLogger("zen.NtpMonitor")
 
-class NtpMonitorDataSource(ZenPackPersistence, RRDDataSource.RRDDataSource):
 
-    ZENPACKID = 'ZenPacks.zenoss.NtpMonitor'
-    NTP_MONITOR = 'NtpMonitor'
-    
-    parser = 'Nagios'
+class NtpMonitorDataSource(PythonDataSource):
```
Comment (StevePC) [inline ZenPacks/zenoss/NtpMonitor/datasources/NtpMonitorDataSource.py]: You'll need to provide a migration for old datasources that are the incorrect type, and you might think about switching from objects.xml to ZPL

Comment (karolzen) [reply ZenPacks/zenoss/NtpMonitor/datasources/NtpMonitorDataSource.py]: I switched to `.yaml` template. Thanks for suggestion.

File: ZenPacks/zenoss/NtpMonitor/ntp.py [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+        self.host = host
+        if port:
+            try:
+                self.port = int(port)
+            except ValueError:
+                log.debug("Wrong value for port is specified. Using default: %i.", self.port)
+        self.version = version
+        if timeout:
+            try:
+                self.timeout = float(timeout)
```
[the remark this answers is not in the document]
Comment (karolzen) [reply ZenPacks/zenoss/NtpMonitor/ntp.py]: I set value for `timeout` at line 205, here I'm checking if argument was passed to `__init__` func.

File: ZenPacks/zenoss/NtpMonitor/ntp.py [hunk 5]
```diff
@@ -0,0 +1,427 @@
...
+        peers_to_check = {}
+        self.sequence_counter += 1
+        packet = NTPPacket(self.version, sequence=self.sequence_counter)
+        more_packets = True
+        while more_packets:
+            self.send_packet(packet.to_data_readstat())
+            log.debug("READSTAT request was sent to host %s", self.host)
+            response_data, src_addr = self.socket.recvfrom(256)
+            log.debug("READSTAT response was received from host %s", self.host)
+            more_packets = False
```
[the remark this answers is not in the document]
Comment (karolzen) [reply ZenPacks/zenoss/NtpMonitor/ntp.py]: Good catch, I left this line here by accident :disappointed:

File: ZenPacks/zenoss/NtpMonitor/ntp.py [hunk 6]
```diff
@@ -0,0 +1,596 @@
...
+        :param data: binary data
+        :return: NtpPacket with extracted data
+        :rtype: NtpPacket
+        """
+        try:
+            unpacked = struct.unpack(
+                NtpPacket._BASE_FIELDS,
+                data[0:struct.calcsize(NtpPacket._BASE_FIELDS)]
+            )
+        except struct.error:
```
Comment (jpeacock-zenoss) [inline ZenPacks/zenoss/NtpMonitor/ntp.py]: Please capture the exception and include its data in the `NtpException`.

File: ZenPacks/zenoss/NtpMonitor/ntp.py [hunk 7]
```diff
@@ -0,0 +1,596 @@
...
+            log.debug("Error during extracting data from NTP packet")
+            raise NtpException("Invalid packet received from NTP server")
```
Comment (jpeacock-zenoss) [inline ZenPacks/zenoss/NtpMonitor/ntp.py]: No need to record this information twice.  The debug log can be removed.

File: ZenPacks/zenoss/NtpMonitor/ntp.py [hunk 8]
```diff
@@ -0,0 +1,596 @@
...
+    def peers(self):
+        """
+        Extract data about peers from NTP packet's data field.
+        Pair of 2 bytes per one peer.
+        """
+        peers = {}
+        if self._peerData:
+            try:
+                unpacked = struct.unpack('!{0}H'.format(self.count / 2), self._peerData)
+            except struct.error:
```
Comment (jpeacock-zenoss) [inline ZenPacks/zenoss/NtpMonitor/ntp.py]: If there's an error with unpacking, it'll be nice to include the details of the error in the `NtpException`.

File: ZenPacks/zenoss/NtpMonitor/ntp.py [hunk 9]
```diff
@@ -0,0 +1,596 @@
...
+                log.debug("Error during extracting data from NTP packet")
+                raise NtpException("Invalid packet received from NTP server")
```
Comment (jpeacock-zenoss) [inline ZenPacks/zenoss/NtpMonitor/ntp.py]: Don't need a debug log if an exception is being raised.

File: ZenPacks/zenoss/NtpMonitor/ntp.py [hunk 10]
```diff
@@ -0,0 +1,596 @@
...
+        self.syncSource = False
+        self.liAlarm = False
+        self.d = None
+        self.timeoutCall = None
+
+    def parsePort(self, port):
+        try:
+            self.port = int(port)
+        except ValueError:
+            log.debug("Wrong value for port is specified. Using default: %i",
```
Comment (jpeacock-zenoss) [inline ZenPacks/zenoss/NtpMonitor/ntp.py]: It'll be helpful to include the value that couldn't be parsed.
Also, this should be a warning log, not debug.

File: ZenPacks/zenoss/NtpMonitor/ntp.py [hunk 11]
```diff
@@ -0,0 +1,596 @@
...
+                      self.port)
+
+    def parseTimeout(self, timeout):
+        try:
+            self.timeout = float(timeout)
+        except ValueError:
+            log.debug("Unable to parse timeout's value. Using default: %.2fs",
```
Comment (jpeacock-zenoss) [inline ZenPacks/zenoss/NtpMonitor/ntp.py]: It'll be helpful to include the value that couldn't be parsed in the log message.
Also, this should be a warning log, not debug.

File: ZenPacks/zenoss/NtpMonitor/ntp.py [hunk 12]
```diff
@@ -0,0 +1,596 @@
...
+        self.errorBit = 0
+        self.data = None
+        self._peerData = None
+
+    def toDataReadstat(self):
+        """
+        Converts data required by READSTAT request into binary format.
+        :return: binary format of NTP READSTAT packet.
+        :rtype: str
+        """
```
[the remark this answers is not in the document]
Comment (karolzen) [reply ZenPacks/zenoss/NtpMonitor/ntp.py]: 👍

File: ZenPacks/zenoss/NtpMonitor/ntp.py [hunk 13]
```diff
@@ -0,0 +1,596 @@
...
+                    self.opcode,
+                    self.sequence,
+                    self.status,
+                    self.assoc,
+                    self.offset,
+                    self.count
+                )
+            except struct.error:
+                log.debug("Problem with parsing packet's data")
+                raise NtpException("Invalid packet received from NTP server")
```
[the remark this answers is not in the document]
Comment (karolzen) [reply ZenPacks/zenoss/NtpMonitor/ntp.py]: 👍

File: ZenPacks/zenoss/NtpMonitor/ntp.py [hunk 14]
```diff
@@ -0,0 +1,596 @@
...
+            raise NtpException(
+                "Version %i of NTP protocol is not implemented" % self.version
+            )
+
+    def toDataReadvar(self):
+        """
+        Converts data required by READVAR request into binary format.
+        :return: binary format of NTP READVAR packet.
+        :rtype: str
+        """
```
[the remark this answers is not in the document]
Comment (karolzen) [reply ZenPacks/zenoss/NtpMonitor/ntp.py]: 👍

File: ZenPacks/zenoss/NtpMonitor/datasources/NtpMonitorDataSource.py [hunk 15]
```diff
@@ -1,114 +1,152 @@
...
+        elif result["liAlarm"]:
+            summary += " Server has the LI_ALARM bit set"
+        if result["offsetResult"] == STATE_UNKNOWN:
+            summary += " Offset unknown"
+        elif result["status"] == STATE_WARNING:
+            summary += " Offset %.10g secs (WARNING)" % result["offset"]
+        elif result["status"] == STATE_CRITICAL:
+            summary += " Offset %.10g secs (CRITICAL)" % result["offset"]
+        else:
+            summary += " Offset %.10g secs" % result["offset"]
```
Comment (jpeacock-zenoss) [inline ZenPacks/zenoss/NtpMonitor/datasources/NtpMonitorDataSource.py]: The rules for building the `summary` string look confusing.  I hope the unit tests all cover the permutations.

File: ZenPacks/zenoss/NtpMonitor/ntp.py [hunk 16]
```diff
@@ -0,0 +1,609 @@
...
+        self.version = version
+        self.opcode = opcode
+        self.sequence = sequence
+        self.status = 0
+        self.mode = 6  # 0x06 - NTP control message
+        self.assoc = 0
+        self.offset = 0
+        self.count = 0
+        self.errorBit = 0
+        self.data = None
```
Comment (jpeacock-zenoss) [inline ZenPacks/zenoss/NtpMonitor/ntp.py]: There's logic around `data`, `count`, and `assoc` that's not encapsulated into the `NtpPacket` class.  These attribute's state appear to split between `NtpPacket` and `NtpProtocol`.

File: ZenPacks/zenoss/NtpMonitor/ntp.py [hunk 17]
```diff
@@ -0,0 +1,609 @@
...
+                    ),
+                    self.opcode,
+                    self.sequence,
+                    self.status,
+                    self.assoc,
+                    self.offset,
+                    self.count
+                )
+                packed += struct.pack(
+                    NtpPacket._DATA.format(len(self.data)), self.data
```
Comment (jpeacock-zenoss) [inline ZenPacks/zenoss/NtpMonitor/ntp.py]: According to `NtpProtocol`, `len(self.data) == self.count`.  `NtpPacket` doesn't seem to be in control of its `data` attribute.

File: ZenPacks/zenoss/NtpMonitor/ntp.py [hunk 18]
```diff
@@ -0,0 +1,609 @@
...
+            self.checkCandidates()
+            self.controlReadvarExchange()
+
+    def sendReadvarRequest(self):
+        log.debug("Getting offset for peer %d", self.currentPeer)
+        packet = NtpPacket(
+            version=self.version, sequence=self.sequenceCounter, opcode=2
+        )
+        packet.assoc = self.currentPeer
+        packet.data = self.getvar
```
Comment (jpeacock-zenoss) [inline ZenPacks/zenoss/NtpMonitor/ntp.py]: So, `data` contains the name of the variable to request?  The variable's name, `data` , seems too generic.

File: ZenPacks/zenoss/NtpMonitor/ntp.py (hunk 16 above)
Comment (karolzen) [reply ZenPacks/zenoss/NtpMonitor/ntp.py]: I'll move this logic to NtpPacket class :+1:

File: ZenPacks/zenoss/NtpMonitor/ntp.py (hunk 18 above)
Comment (karolzen) [reply ZenPacks/zenoss/NtpMonitor/ntp.py]: Field which contains details about requested value in NTP's packet structure is called "Data"

File: ZenPacks/zenoss/NtpMonitor/datasources/NtpMonitorDataSource.py [hunk 19]
```diff
@@ -1,114 +1,152 @@
...
+    Datasource for NTP protocol.
+    """
+    ZENPACKID = "ZenPacks.zenoss.NtpMonitor"
+    NTP_MONITOR = "NtpMonitor"
 
     sourcetypes = (NTP_MONITOR,)
     sourcetype = NTP_MONITOR
 
+    plugin_classname = "ZenPacks.zenoss.NtpMonitor.datasources." \
+                       "NtpMonitorDataSource.NtpMonitorDataSourcePlugin"
```
[the remark this answers is not in the document]
Comment (karolzen) [reply ZenPacks/zenoss/NtpMonitor/datasources/NtpMonitorDataSource.py]: Haha, looks better for me. Updated :+1:

File: ZenPacks/zenoss/NtpMonitor/__init__.py [hunk 20]
```diff
@@ -19,3 +28,6 @@ def onCollectorInstalled(ob, event):
     code, output = ob.executeCommand('zenbincheck %s' % verifyBin, 'zenoss', needsZenHome=True)
     if code:
        	log.warn(errormsg.format(verifyBin, ob.hostname, zpFriendly))
```
Comment (jpeacock-zenoss) [inline ZenPacks/zenoss/NtpMonitor/__init__.py]: There's a tab in this whitespace.

**Pull request merged**